src/transaction/log_page_buffer.c

```diff
@@ -1123,123 +771,85 @@ logpb_fix_page (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, PAGE_FETCH_MODE fetc
   assert (LOG_CS_OWN_WRITE_MODE (thread_p));
 
-  START_EXCLUSIVE_ACCESS_LOG_PB (rv, thread_p);
-
-  log_bufptr = (LOG_BUFFER *) mht_get (log_Pb.ht, &pageid);
-  if (log_bufptr == NULL)
+  if (pageid == LOGPB_HEADER_PAGE_ID)
     {
-      /* 
-       * Page is not resident. Find a replacement buffer for it
-       */
-
-      stat_page_found = PERF_PAGE_MODE_OLD_LOCK_WAIT;
-
-      while (1)
+      log_bufptr = &log_Pb.header_buffer;
+    }
+  else
+    {
+      index = logpb_get_log_buffer_index ((int) pageid);
+      if (index >= 0 && index < log_Pb.num_buffers)
 	{
-	  log_bufptr = logpb_replace (thread_p, &retry);
-	  if (log_bufptr == NULL)
-	    {
-	      if (retry)
-		{
-		  continue;
-		}
-	      else
-		{
-		  goto error;
-		}
-	    }
-	  else
-	    {
-	      break;
-	    }
+	  log_bufptr = &log_Pb.buffers[index];
 	}
-
-      assert (log_bufptr != NULL);
-
-      /* 
-       * Remove the current page from the hash table and get the desired
-       * page from disk when the page is an old page
-       */
-      if (log_bufptr->pageid != NULL_PAGEID)
+      else
 	{
-	  (void) mht_rem (log_Pb.ht, &log_bufptr->pageid, NULL, NULL);
+	  er_set (ER_FATAL_ERROR_SEVERITY, ARG_FILE_LINE, ER_LOG_PAGE_CORRUPTED, 1, pageid);
+	  return NULL;
 	}
 
-      /* Fix the page and mark its pageid invalid */
-      log_bufptr->pageid = NULL_PAGEID;
-      log_bufptr->fcnt++;
+    }
+  assert (log_bufptr != NULL);
 
-      phy_pageid = logpb_to_physical_pageid (pageid);
+  if (log_bufptr->pageid != NULL_PAGEID && log_bufptr->pageid != pageid)
+    {
+      if (log_bufptr->dirty == true)
+	{
+	  if (logpb_write_page_to_disk (thread_p, log_bufptr->logpage, log_bufptr->pageid) != NO_ERROR)
+	    {
+	      assert_release (false);
+	      return NULL;
+	    }
+	  log_bufptr->dirty = false;
+	}
+      log_bufptr->pageid = NULL_PAGEID;	/* invalidate buffer */
+      mnt_log_replacements (thread_p);
+    }
 
+  if (log_bufptr->pageid == NULL_PAGEID)
+    {
       if (fetch_mode == NEW_PAGE)
 	{
-	  log_bufptr->logpage.hdr.logical_pageid = pageid;
-	  log_bufptr->logpage.hdr.offset = NULL_OFFSET;
+	  log_bufptr->logpage->hdr.logical_pageid = pageid;
+	  log_bufptr->logpage->hdr.offset = NULL_OFFSET;
 	}
       else
 	{
-	  if (logpb_read_page_from_file (thread_p, pageid, LOG_CS_FORCE_USE, &log_bufptr->logpage) == NULL)
+	  stat_page_found = PERF_PAGE_MODE_OLD_LOCK_WAIT;
+	  if (logpb_read_page_from_file (thread_p, pageid, LOG_CS_FORCE_USE, log_bufptr->logpage) != NO_ERROR)
 	    {
-	      goto error;
+	      assert_release (false);
+	      return NULL;
 	    }
-	  mnt_log_fetch_ioreads (thread_p);
-	}
 
-      /* Recall the page in the buffer pool, and hash the identifier */
-      log_bufptr->pageid = pageid;
-      log_bufptr->phy_pageid = phy_pageid;
-
-      if (mht_put (log_Pb.ht, &log_bufptr->pageid, log_bufptr) == NULL)
-	{
-	  logpb_initialize_log_buffer (log_bufptr);
-	  log_bufptr = NULL;
+	  mnt_log_ioreads (thread_p);
```

Comment:
@funkydvd 

Please notice that `logpb_read_page_from_file` counts the number of log reads and remove `mnt_log_ioreads` call from here. :)